pkg/cloudprovider/vsphere/nodemanager_test.go

```diff
@@ -267,3 +268,1058 @@ func TestReturnIPsFromSpecificFamily(t *testing.T) {
+				{Type: "ExternalIP", Address: "172.15.108.12"},
+			},
+		},
+		{
+			testName: "DualStack_whenNoIPsOfOneFamilyAreDiscovered",
+			setup: testSetup{
+				ipFamilyPriority: []string{"ipv6", "ipv4"},
+				cpiConfig:        nil,
+				networks: []vimtypes.GuestNicInfo{
+					{
+						Network: "internal_net",
+						IpAddress: []string{
+							"127.0.0.1",
+							"fd00:cccc::1",
+						},
+					},
+				},
+			},
+			expectedErrorSubstring: "unable to find suitable IP address for node",
+		},
+	}
+
+	for _, testcase := range testcases {
+		t.Run(testcase.testName, func(t *testing.T) {
+			cfg, fin := configFromEnvOrSim(true)
+			defer fin()
+
+			cfg.VirtualCenter[cfg.Global.VCenterIP].IPFamilyPriority = testcase.setup.ipFamilyPriority
+			connMgr := cm.NewConnectionManager(cfg, nil, nil)
+			defer connMgr.Logout()
+
+			nm := newNodeManager(testcase.setup.cpiConfig, connMgr)
+
+			vm := simulator.Map.Any("VirtualMachine").(*simulator.VirtualMachine)
+			vm.Guest.HostName = strings.ToLower(vm.Name) // simulator.SearchIndex.FindByDnsName matches against the guest.hostName property
+			vm.Guest.Net = testcase.setup.networks
+
+			name := vm.Name
+
+			err := connMgr.Connect(context.Background(), connMgr.VsphereInstanceMap[cfg.Global.VCenterIP])
+			if err != nil {
+				t.Errorf("Failed to Connect to vSphere: %s", err)
+			}
+
+			// subject
+			err = nm.DiscoverNode(name, cm.FindVMByName)
+			if testcase.expectedErrorSubstring != "" {
+				if err == nil {
+					t.Errorf("failed: expected DiscoverNode to return error containing: %q but no error occurred", testcase.expectedErrorSubstring)
+					return
+				}
+				if !strings.Contains(err.Error(), testcase.expectedErrorSubstring) {
+					t.Errorf("failed: expected DiscoverNode to return error containing: %q but was %q", testcase.expectedErrorSubstring, err.Error())
+				}
+				return
+			} else if err != nil {
+				t.Errorf("Failed DiscoverNode: %s", err)
+				return
+			}
+
+			nodeInfo, ok := nm.nodeNameMap[strings.ToLower(name)]
+			if !ok {
+				t.Errorf("failed: %v not found", name)
+			}
+
+			// hostname is always returned first, then the expected ips
+			expectations := append(
+				[]v1.NodeAddress{{Type: "Hostname", Address: strings.ToLower(vm.Name)}},
+				testcase.expectedIPs...,
+			)
+			if len(nodeInfo.NodeAddresses) != len(expectations) {
+				t.Errorf("failed: nodeInfo.NodeAddresses should be length %d but was %d", len(testcase.expectedIPs)+1, len(nodeInfo.NodeAddresses))
+			}
+			for i, nodeAddress := range expectations {
+				if nodeInfo.NodeAddresses[i].Address != nodeAddress.Address {
+					t.Errorf("failed: NodeAddresses[%d].Address should eq %q but was %q", i, nodeAddress.Address, nodeInfo.NodeAddresses[i].Address)
+				}
+				if nodeInfo.NodeAddresses[i].Type != nodeAddress.Type {
+					t.Errorf("failed: NodeAddresses[%d].Type should eq %q but was %q", i, nodeAddress.Type, nodeInfo.NodeAddresses[i].Type)
+				}
+			}
+		})
+	}
+}
```

Comment:
would it be possible to add unit tests for every new function we introduce in nodemanager.go?

Reply:
It's our preference to test nodemanager as a black box, testing inputs and outputs. We'd like to test behavior and not implementation. That said, if you have a strong preference, we can test the private funcs. We'll leave it up to you.

Reply:
I feel it's better to have unit tests for private funcs but it's totally okay to leave them as follow-up PRs